File: src/lib/parser.ts
import { getDb } from './db/index';
import { localIsoDate } from './images';

export type ParsedItem =
  | { type: 'role'; role: string; name: string; timing?: string | null }
  | {
      type: 'hymn';
      number: number;
      title: string;
      lyrics: string;
      incomplete?: boolean;
      timing?: string | null;
    }
  | { type: 'section'; title: string; timing?: string | null };

export interface ParsedSermon {
  speaker: string;
  title: string;
}

/** Sender-supplied scripture text (theme / verse reading). */
export interface ParsedScripture {
  reference: string | null;
  text: string;
  /** Corpus that produced `text` when the operator resolved the reference. */
  translation?: string;
}

export interface ParsedRundown {
  date: string | null;
  items: ParsedItem[];
  unmappedLines: string[];
  /** SDAH numbers that failed hymnal lookup (FR-2). */
  failedHymnNumbers: number[];
  sermon: ParsedSermon | null;
  /** null/empty when Special Song is `-` (none). */
  specialSong: string | null;
  closingPrayerPerson: string | null;
  /** null → PPTX uses standing default theme verse. */
  themeVerse: ParsedScripture | null;
  /** null → omit Verse Reading slide. */
  verseReading: ParsedScripture | null;
  /** Freeform family/youth-of-the-week text; null → omit slide. Legacy combined field. */
  familyYouth: string | null;
  /** Family-of-the-week prayer request (Slide 56). */
  familyPrayerRequest: string | null;
  /** Youth-of-the-week prayer request (Slide 56). */
  youthPrayerRequest: string | null;
}

function normalizeNewlines(text: string): string {
  return text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
}

function parseCalendarDate(rawDate: string): string | null {
  if (/^\d{4}-\d{2}-\d{2}$/.test(rawDate)) {
    const [y, m, d] = rawDate.split('-').map(Number);
    const dt = new Date(Date.UTC(y, m - 1, d));
    if (
      dt.getUTCFullYear() !== y ||
      dt.getUTCMonth() !== m - 1 ||
      dt.getUTCDate() !== d
    ) {
      return null;
    }
    return rawDate;
  }

  const d = new Date(rawDate);
  if (isNaN(d.getTime())) return null;
  return localIsoDate(d);
}

/** Strip addendum markers `》` and `[  ]`. */
function stripPrefixes(line: string): string {
  return line
    .replace(/^》\s*/, '')
    .replace(/^\[\s*\]\s*/, '')
    .trim();
}

const TIMING_RANGE_RE =
  /\(\s*\d{1,2}[.:]\d{2}\s*[-–]\s*\d{1,2}[.:]\d{2}\s*\/?\s*\d*\s*min?\s*\)/gi;
const TIMING_MINUTES_RE = /\(\s*\d+\s*min(?:ute)?s?\s*\)/gi;
const TIMING_M_RE = /\(\s*\d+\s*m\s*\)/gi;

/** Extract duration timings like `(5m)`, `(45m)`, and section time ranges. */
export function extractTiming(line: string): string | null {
  const found: string[] = [];
  for (const re of [TIMING_RANGE_RE, TIMING_MINUTES_RE, TIMING_M_RE]) {
    re.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = re.exec(line)) !== null) {
      found.push(m[0].replace(/^\(|\)$/g, '').trim());
    }
  }
  if (found.length === 0) return null;
  return found.join(' · ');
}

/** Strip duration timings like `(5m)`, `(45m)`, and section time ranges. */
function stripTimings(line: string): string {
  return line
    .replace(TIMING_RANGE_RE, '')
    .replace(TIMING_MINUTES_RE, '')
    .replace(TIMING_M_RE, '')
    .replace(/\s{2,}/g, ' ')
    .trim();
}

/** Strip prefixes; keep timings for run-sheet (use stripTimings for role matching). */
function cleanLine(line: string): string {
  return stripTimings(stripPrefixes(line));
}

function withTiming<T extends ParsedItem>(item: T, timing: string | null): T {
  if (!timing) return item;
  return { ...item, timing };
}

function isSectionHeader(line: string): boolean {
  return /^(BIBLE\s+TALK|DIVINE\s+SERVICE|BREAK)\b/i.test(line);
}

function parseSermonLine(
  line: string
): ParsedSermon | null {
  const m = line.match(
    /^Sermon\s*[:\-]\s*(.+?)(?:\s+"([^"]+)"|\s+[“"]([^”"]+)[”"])?\s*$/i
  );
  if (!m) return null;
  const speaker = m[1].replace(/\s+"[^"]*"\s*$/, '').trim();
  const title = (m[2] || m[3] || '').trim();
  if (!speaker) return null;
  return { speaker, title };
}

function parseSpecialSong(line: string): string | null | undefined {
  const m = line.match(/^Special\s+Song\s*[:\-]\s*(.*)$/i);
  if (!m) return undefined;
  const value = m[1].trim();
  if (!value || value === '-' || value === '—' || /^none$/i.test(value)) {
    return null;
  }
  return value;
}

/** Split "Acts 18:9,10 — text" / "John 4:23: text" into reference + text. */
export function parseScriptureValue(raw: string): ParsedScripture | null {
  const value = raw.trim();
  if (!value || value === '-' || value === '—') return null;

  const split = value.match(
    /^(.+?)\s+(\d+:[\d,\-–]+)(?:\s*[—–\-:]\s*|\s+)(.+)$/
  );
  if (split) {
    return {
      reference: `${split[1].trim()} ${split[2].trim()}`,
      text: split[3].trim(),
    };
  }

  const refOnly = value.match(
    /^(.+?)\s+(\d+:[\d,\-–]+)\s*$/
  );
  if (refOnly) {
    return {
      reference: `${refOnly[1].trim()} ${refOnly[2].trim()}`,
      text: '',
    };
  }

  return { reference: null, text: value };
}

function parseThemeVerse(line: string): ParsedScripture | null | undefined {
  const m = line.match(/^Theme(?:\s+Verse)?\s*[:\-]\s*(.*)$/i);
  if (!m) return undefined;
  return parseScriptureValue(m[1]);
}

function parseVerseReading(line: string): ParsedScripture | null | undefined {
  const m = line.match(
    /^(?:Verse\s+Reading|Memory\s+(?:Verse|Text)|Ayat\s+Bacaan)\s*[:\-]\s*(.*)$/i
  );
  if (!m) return undefined;
  return parseScriptureValue(m[1]);
}

function parseFamilyYouth(line: string): string | null | undefined {
  const m = line.match(
    /^(?:Family(?:\s*&\s*|\s+and\s+|\/\s*)Youth(?:\s+of\s+the\s+Week)?|Family\s+of\s+the\s+Week|Youth\s+of\s+the\s+Week|Keluarga(?:\s*&\s*|\s+dan\s+)Pemuda)\s*[:\-]\s*(.*)$/i
  );
  if (!m) return undefined;
  const value = m[1].trim();
  if (!value || value === '-' || value === '—') return null;
  return value;
}

function parseRoleLine(line: string): { role: string; name: string } | null {
  // Avoid treating hymn/sermon/special-song lines as roles
  if (/(?:SDAH|Hymn|#)\s*\d+/i.test(line)) return null;
  if (/^Sermon\s*[:\-]/i.test(line)) return null;
  if (/^Special\s+Song\s*[:\-]/i.test(line)) return null;
  if (/^Theme(?:\s+Verse)?\s*[:\-]/i.test(line)) return null;
  if (
    /^(?:Verse\s+Reading|Memory\s+(?:Verse|Text)|Ayat\s+Bacaan)\s*[:\-]/i.test(
      line
    )
  ) {
    return null;
  }
  if (
    /^(?:Family(?:\s*&\s*|\s+and\s+|\/\s*)Youth|Family\s+of\s+the\s+Week|Youth\s+of\s+the\s+Week|Keluarga)/i.test(
      line
    )
  ) {
    return null;
  }
  if (isSectionHeader(line) && !line.includes(':')) return null;

  const bracket = line.match(/^\[([^\]]+)\]\s*(.+)$/);
  const m = bracket || line.match(/^(.+?)\s*[:\-]\s*(.+)$/);
  if (!m) return null;

  const role = m[1].trim();
  const name = m[2].trim();
  if (!role || !name) return null;
  if (/^\d{1,2}:\d{2}$/.test(role)) return null;

  return { role, name };
}

/** Resolve an SDAH number from the hymns table (lyrics + incomplete flag). */
export function lookupHymn(number: number): {
  title: string;
  lyrics: string;
  incomplete?: boolean;
} {
  const db = getDb();
  const hymnRecord = db
    .prepare('SELECT title, lyrics FROM hymns WHERE number = ?')
    .get(number) as { title: string; lyrics: string } | undefined;

  if (hymnRecord) {
    if (!hymnRecord.lyrics?.trim()) {
      return {
        title: hymnRecord.title || `Unknown SDAH ${number}`,
        lyrics: '',
        incomplete: true,
      };
    }
    return { title: hymnRecord.title, lyrics: hymnRecord.lyrics };
  }

  return {
    title: `Unknown SDAH ${number}`,
    lyrics: '',
    incomplete: true,
  };
}

export function parseRundown(rawText: string): ParsedRundown {
  const normalizedText = normalizeNewlines(rawText);
  const lines = normalizedText
    .split('\n')
    .map((l) => l.trim())
    .filter((l) => l.length > 0);

  const parsed: ParsedRundown = {
    date: null,
    items: [],
    unmappedLines: [],
    failedHymnNumbers: [],
    sermon: null,
    specialSong: null,
    closingPrayerPerson: null,
    themeVerse: null,
    verseReading: null,
    familyYouth: null,
    familyPrayerRequest: null,
    youthPrayerRequest: null,
  };

  const dateRegex =
    /(?:20\d{2}-\d{2}-\d{2})|(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\s+\d{1,2},?\s+20\d{2}/i;
  const hymnRegex = /(?:SDAH|Hymn|#)\s*(\d+)/i;

  const dateMatch = normalizedText.match(dateRegex);
  if (dateMatch) {
    parsed.date = parseCalendarDate(dateMatch[0]);
  }

  let sermonSpeaker: string | null = null;

  for (const rawLine of lines) {
    if (rawLine.match(dateRegex) && !rawLine.includes(':')) {
      continue;
    }

    const timing = extractTiming(stripPrefixes(rawLine));
    const line = cleanLine(rawLine);
    if (!line) continue;

    let mapped = false;

    // Section headers (BIBLE TALK / DIVINE SERVICE / Break)
    if (isSectionHeader(line)) {
      const title = line
        .replace(/\s*\(.*\)\s*$/, '')
        .replace(/\s+/g, ' ')
        .trim();
      parsed.items.push(withTiming({ type: 'section', title }, timing));
      mapped = true;
    }

    // Special Song
    if (!mapped) {
      const special = parseSpecialSong(line);
      if (special !== undefined) {
        parsed.specialSong = special;
        mapped = true;
      }
    }

    // Theme verse (optional; standing default applied at PPTX time)
    if (!mapped) {
      const theme = parseThemeVerse(line);
      if (theme !== undefined) {
        parsed.themeVerse = theme;
        mapped = true;
      }
    }

    // Verse Reading / Memory Verse
    if (!mapped) {
      const verse = parseVerseReading(line);
      if (verse !== undefined) {
        parsed.verseReading = verse;
        mapped = true;
      }
    }

    // Family & Youth of the Week
    if (!mapped) {
      const family = parseFamilyYouth(line);
      if (family !== undefined) {
        parsed.familyYouth = family;
        mapped = true;
      }
    }

    // Sermon
    if (!mapped) {
      const sermon = parseSermonLine(line);
      if (sermon) {
        parsed.sermon = sermon;
        sermonSpeaker = sermon.speaker;
        parsed.items.push(
          withTiming(
            {
              type: 'role' as const,
              role: 'Sermon',
              name: sermon.title
                ? `${sermon.speaker} — ${sermon.title}`
                : sermon.speaker,
            },
            timing
          )
        );
        mapped = true;
      }
    }

    // Hymns (prefer DB title over rundown trailing text)
    if (!mapped) {
      const hymnMatch = line.match(hymnRegex);
      if (hymnMatch) {
        const number = parseInt(hymnMatch[1], 10);
        const found = lookupHymn(number);
        parsed.items.push(
          withTiming(
            {
              type: 'hymn' as const,
              number,
              title: found.title,
              lyrics: found.lyrics,
              ...(found.incomplete ? { incomplete: true as const } : {}),
            },
            timing
          )
        );
        if (found.incomplete && !parsed.failedHymnNumbers.includes(number)) {
          parsed.failedHymnNumbers.push(number);
        }
        mapped = true;
      }
    }

    // Roles (incl. Closing Prayer → The Speaker)
    if (!mapped) {
      const role = parseRoleLine(line);
      if (role) {
        let name = role.name;
        if (
          /^Closing\s+Prayer$/i.test(role.role) &&
          /^The\s+Speaker$/i.test(name)
        ) {
          name = sermonSpeaker || name;
        }
        if (/^Closing\s+Prayer$/i.test(role.role)) {
          parsed.closingPrayerPerson = name;
        }
        parsed.items.push(
          withTiming({ type: 'role' as const, role: role.role, name }, timing)
        );
        mapped = true;
      }
    }

    if (!mapped && line.match(dateRegex)) {
      mapped = true;
    }

    if (!mapped) {
      parsed.unmappedLines.push(rawLine);
    }
  }

  // If Closing Prayer used The Speaker before sermon was seen (unlikely),
  // or was set to literal, re-resolve from structured sermon.
  if (
    parsed.closingPrayerPerson &&
    /^The\s+Speaker$/i.test(parsed.closingPrayerPerson) &&
    sermonSpeaker
  ) {
    parsed.closingPrayerPerson = sermonSpeaker;
    for (const item of parsed.items) {
      if (
        item.type === 'role' &&
        /^Closing\s+Prayer$/i.test(item.role) &&
        /^The\s+Speaker$/i.test(item.name)
      ) {
        item.name = sermonSpeaker;
      }
    }
  }

  return parsed;
}

File: internal/parse/parser.go
package parse

import (
	"database/sql"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

type Scripture struct {
	Reference   *string `json:"reference"`
	Text        string  `json:"text"`
	Translation string  `json:"translation,omitempty"`
}

type Sermon struct {
	Speaker string `json:"speaker"`
	Title   string `json:"title"`
}

type Item struct {
	Type       string  `json:"type"`
	Role       string  `json:"role,omitempty"`
	Name       string  `json:"name,omitempty"`
	Title      string  `json:"title,omitempty"`
	Number     int     `json:"number,omitempty"`
	Lyrics     string  `json:"lyrics,omitempty"`
	Incomplete bool    `json:"incomplete,omitempty"`
	Timing     *string `json:"timing,omitempty"`
}

type Rundown struct {
	Date                *string    `json:"date"`
	Items               []Item     `json:"items"`
	UnmappedLines       []string   `json:"unmappedLines"`
	FailedHymnNumbers   []int      `json:"failedHymnNumbers"`
	Sermon              *Sermon    `json:"sermon"`
	SpecialSong         *string    `json:"specialSong"`
	ClosingPrayerPerson *string    `json:"closingPrayerPerson"`
	ThemeVerse          *Scripture `json:"themeVerse"`
	VerseReading        *Scripture `json:"verseReading"`
	FamilyYouth         *string    `json:"familyYouth"`
	FamilyPrayerRequest *string    `json:"familyPrayerRequest"`
	YouthPrayerRequest  *string    `json:"youthPrayerRequest"`
}

var (
	timingRange    = regexp.MustCompile(`(?i)\(\s*\d{1,2}[.:]\d{2}\s*[-–]\s*\d{1,2}[.:]\d{2}\s*/?\s*\d*\s*min?\s*\)`)
	timingMinutes  = regexp.MustCompile(`(?i)\(\s*\d+\s*min(?:ute)?s?\s*\)`)
	timingM        = regexp.MustCompile(`(?i)\(\s*\d+\s*m\s*\)`)
	dateRE         = regexp.MustCompile(`(?i)(?:20\d{2}-\d{2}-\d{2})|(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\s+\d{1,2},?\s+20\d{2}`)
	hymnRE         = regexp.MustCompile(`(?i)(?:SDAH|Hymn|#)\s*(\d+)`)
	sectionRE      = regexp.MustCompile(`(?i)^(BIBLE\s+TALK|DIVINE\s+SERVICE|BREAK)\b`)
	sermonRE       = regexp.MustCompile(`(?i)^Sermon\s*[:\-]\s*(.+?)(?:\s+"([^"]+)"|\s+[“"]([^”"]+)[”"])?\s*$`)
	specialRE      = regexp.MustCompile(`(?i)^Special\s+Song\s*[:\-]\s*(.*)$`)
	themeRE        = regexp.MustCompile(`(?i)^Theme(?:\s+Verse)?\s*[:\-]\s*(.*)$`)
	verseRE        = regexp.MustCompile(`(?i)^(?:Verse\s+Reading|Memory\s+(?:Verse|Text)|Ayat\s+Bacaan)\s*[:\-]\s*(.*)$`)
	familyRE       = regexp.MustCompile(`(?i)^(?:Family(?:\s*&\s*|\s+and\s+|/\s*)Youth(?:\s+of\s+the\s+Week)?|Family\s+of\s+the\s+Week|Youth\s+of\s+the\s+Week|Keluarga(?:\s*&\s*|\s+dan\s+)Pemuda)\s*[:\-]\s*(.*)$`)
	hymnHintRE     = regexp.MustCompile(`(?i)(?:SDAH|Hymn|#)\s*\d+`)
	scriptureSplit = regexp.MustCompile(`^(.+?)\s+(\d+:[\d,\-–]+)(?:\s*[—–\-:]\s*|\s+)(.+)$`)
	scriptureRef   = regexp.MustCompile(`^(.+?)\s+(\d+:[\d,\-–]+)\s*$`)
	bracketRole    = regexp.MustCompile(`^\[([^\]]+)\]\s*(.+)$`)
	colonRole      = regexp.MustCompile(`^(.+?)\s*[:\-]\s*(.+)$`)
	clockRole      = regexp.MustCompile(`^\d{1,2}:\d{2}$`)
	isoDate        = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)
)

func LocalISODate(t time.Time) string {
	return t.Format("2006-01-02")
}

func parseCalendarDate(raw string) *string {
	raw = strings.TrimSpace(raw)
	if isoDate.MatchString(raw) {
		t, err := time.Parse("2006-01-02", raw)
		if err != nil || t.Format("2006-01-02") != raw {
			return nil
		}
		return &raw
	}
	layouts := []string{
		"January 2, 2006",
		"January 2 2006",
		"Jan 2, 2006",
		"Jan 2 2006",
	}
	title := toTitleMonth(raw)
	for _, layout := range layouts {
		if t, err := time.Parse(layout, title); err == nil {
			s := t.Format("2006-01-02")
			return &s
		}
	}
	return nil
}

func toTitleMonth(s string) string {
	parts := strings.Fields(s)
	if len(parts) == 0 {
		return s
	}
	m := strings.ToLower(parts[0])
	if len(m) > 0 {
		parts[0] = strings.ToUpper(m[:1]) + m[1:]
	}
	return strings.Join(parts, " ")
}

func extractTiming(line string) *string {
	var found []string
	for _, re := range []*regexp.Regexp{timingRange, timingMinutes, timingM} {
		for _, m := range re.FindAllString(line, -1) {
			m = strings.TrimPrefix(m, "(")
			m = strings.TrimSuffix(m, ")")
			found = append(found, strings.TrimSpace(m))
		}
	}
	if len(found) == 0 {
		return nil
	}
	joined := strings.Join(found, " · ")
	return &joined
}

func stripTimings(line string) string {
	line = timingRange.ReplaceAllString(line, "")
	line = timingMinutes.ReplaceAllString(line, "")
	line = timingM.ReplaceAllString(line, "")
	return strings.Join(strings.Fields(line), " ")
}

func stripPrefixes(line string) string {
	line = strings.TrimSpace(line)
	line = strings.TrimPrefix(line, "》")
	line = strings.TrimSpace(line)
	if strings.HasPrefix(line, "[") {
		line = regexp.MustCompile(`^\[\s*\]\s*`).ReplaceAllString(line, "")
	}
	return strings.TrimSpace(line)
}

func cleanLine(line string) string {
	return stripTimings(stripPrefixes(line))
}

// resolveDefaultBook resolves the song book code from the database using the
// DEC-004 S3 three-step fallback order:
//  1. explicit weekly/provided bookCode
//  2. global default book in song_books (is_default = 1)
//  3. shipped DefaultSongBook ("SDAH")
func resolveDefaultBook(db *sql.DB, explicitBook string) string {
	explicit := strings.ToUpper(strings.TrimSpace(explicitBook))
	if explicit != "" {
		return explicit
	}
	if db != nil {
		var defaultBook string
		err := db.QueryRow(`SELECT book_code FROM song_books WHERE is_default = 1 LIMIT 1`).Scan(&defaultBook)
		if err == nil && strings.TrimSpace(defaultBook) != "" {
			return strings.ToUpper(strings.TrimSpace(defaultBook))
		}
	}
	return "SDAH"
}

// LookupHymnInBook resolves a hymn on the pair (book_code, number) following the
// DEC-004 S3 fallback order for book resolution.
func LookupHymnInBook(db *sql.DB, bookCode string, number int) (title, lyrics string, incomplete bool) {
	resolvedBook := resolveDefaultBook(db, bookCode)
	if db == nil {
		return fmt.Sprintf("Unknown %s %d", resolvedBook, number), "", true
	}
	var t, l sql.NullString
	err := db.QueryRow(`SELECT title, lyrics FROM hymns WHERE book_code = ? AND number = ?`, resolvedBook, number).Scan(&t, &l)
	if err != nil {
		return fmt.Sprintf("Unknown %s %d", resolvedBook, number), "", true
	}
	title = t.String
	lyrics = l.String
	if strings.TrimSpace(lyrics) == "" {
		if title == "" {
			title = fmt.Sprintf("Unknown %s %d", resolvedBook, number)
		}
		return title, "", true
	}
	return title, lyrics, false
}

// LookupHymn resolves a hymn by number using the default resolved book.
func LookupHymn(db *sql.DB, number int) (title, lyrics string, incomplete bool) {
	return LookupHymnInBook(db, "", number)
}

func ParseScriptureValue(raw string) *Scripture {
	value := strings.TrimSpace(raw)
	if value == "" || value == "-" || value == "—" {
		return nil
	}
	if m := scriptureSplit.FindStringSubmatch(value); m != nil {
		ref := strings.TrimSpace(m[1]) + " " + strings.TrimSpace(m[2])
		return &Scripture{Reference: &ref, Text: strings.TrimSpace(m[3])}
	}
	if m := scriptureRef.FindStringSubmatch(value); m != nil {
		ref := strings.TrimSpace(m[1]) + " " + strings.TrimSpace(m[2])
		return &Scripture{Reference: &ref, Text: ""}
	}
	return &Scripture{Reference: nil, Text: value}
}

func ParseRundown(db *sql.DB, rawText string) Rundown {
	normalized := strings.ReplaceAll(strings.ReplaceAll(rawText, "\r\n", "\n"), "\r", "\n")
	var lines []string
	for _, l := range strings.Split(normalized, "\n") {
		l = strings.TrimSpace(l)
		if l != "" {
			lines = append(lines, l)
		}
	}
	parsed := Rundown{
		Items:             []Item{},
		UnmappedLines:     []string{},
		FailedHymnNumbers: []int{},
	}
	if m := dateRE.FindString(normalized); m != "" {
		parsed.Date = parseCalendarDate(m)
	}
	var sermonSpeaker string
	for _, rawLine := range lines {
		if dateRE.MatchString(rawLine) && !strings.Contains(rawLine, ":") {
			continue
		}
		timing := extractTiming(stripPrefixes(rawLine))
		line := cleanLine(rawLine)
		if line == "" {
			continue
		}
		mapped := false
		if sectionRE.MatchString(line) {
			title := strings.TrimSpace(regexp.MustCompile(`\s*\(.*\)\s*$`).ReplaceAllString(line, ""))
			title = strings.Join(strings.Fields(title), " ")
			parsed.Items = append(parsed.Items, withTiming(Item{Type: "section", Title: title}, timing))
			mapped = true
		}
		if !mapped {
			if m := specialRE.FindStringSubmatch(line); m != nil {
				v := strings.TrimSpace(m[1])
				if v == "" || v == "-" || v == "—" || strings.EqualFold(v, "none") {
					parsed.SpecialSong = nil
				} else {
					parsed.SpecialSong = &v
				}
				mapped = true
			}
		}
		if !mapped {
			if m := themeRE.FindStringSubmatch(line); m != nil {
				parsed.ThemeVerse = ParseScriptureValue(m[1])
				mapped = true
			}
		}
		if !mapped {
			if m := verseRE.FindStringSubmatch(line); m != nil {
				parsed.VerseReading = ParseScriptureValue(m[1])
				mapped = true
			}
		}
		if !mapped {
			if m := familyRE.FindStringSubmatch(line); m != nil {
				v := strings.TrimSpace(m[1])
				if v == "" || v == "-" || v == "—" {
					parsed.FamilyYouth = nil
				} else {
					parsed.FamilyYouth = &v
				}
				mapped = true
			}
		}
		if !mapped {
			if m := sermonRE.FindStringSubmatch(line); m != nil {
				speaker := strings.TrimSpace(regexp.MustCompile(`\s+"[^"]*"\s*$`).ReplaceAllString(m[1], ""))
				title := ""
				if m[2] != "" {
					title = strings.TrimSpace(m[2])
				} else if m[3] != "" {
					title = strings.TrimSpace(m[3])
				}
				if speaker != "" {
					parsed.Sermon = &Sermon{Speaker: speaker, Title: title}
					sermonSpeaker = speaker
					name := speaker
					if title != "" {
						name = speaker + " — " + title
					}
					parsed.Items = append(parsed.Items, withTiming(Item{Type: "role", Role: "Sermon", Name: name}, timing))
					mapped = true
				}
			}
		}
		if !mapped {
			if m := hymnRE.FindStringSubmatch(line); m != nil {
				number, _ := strconv.Atoi(m[1])
				title, lyrics, incomplete := LookupHymn(db, number)
				item := Item{Type: "hymn", Number: number, Title: title, Lyrics: lyrics, Incomplete: incomplete}
				parsed.Items = append(parsed.Items, withTiming(item, timing))
				if incomplete && !containsInt(parsed.FailedHymnNumbers, number) {
					parsed.FailedHymnNumbers = append(parsed.FailedHymnNumbers, number)
				}
				mapped = true
			}
		}
		if !mapped {
			if role := parseRoleLine(line); role != nil {
				name := role.Name
				if regexp.MustCompile(`(?i)^Closing\s+Prayer$`).MatchString(role.Role) &&
					regexp.MustCompile(`(?i)^The\s+Speaker$`).MatchString(name) {
					if sermonSpeaker != "" {
						name = sermonSpeaker
					}
				}
				if regexp.MustCompile(`(?i)^Closing\s+Prayer$`).MatchString(role.Role) {
					parsed.ClosingPrayerPerson = &name
				}
				parsed.Items = append(parsed.Items, withTiming(Item{Type: "role", Role: role.Role, Name: name}, timing))
				mapped = true
			}
		}
		if !mapped && dateRE.MatchString(line) {
			mapped = true
		}
		if !mapped {
			parsed.UnmappedLines = append(parsed.UnmappedLines, rawLine)
		}
	}
	if parsed.ClosingPrayerPerson != nil &&
		regexp.MustCompile(`(?i)^The\s+Speaker$`).MatchString(*parsed.ClosingPrayerPerson) &&
		sermonSpeaker != "" {
		parsed.ClosingPrayerPerson = &sermonSpeaker
		for i, item := range parsed.Items {
			if item.Type == "role" &&
				regexp.MustCompile(`(?i)^Closing\s+Prayer$`).MatchString(item.Role) &&
				regexp.MustCompile(`(?i)^The\s+Speaker$`).MatchString(item.Name) {
				parsed.Items[i].Name = sermonSpeaker
			}
		}
	}
	return parsed
}

type roleLine struct {
	Role string
	Name string
}

func parseRoleLine(line string) *roleLine {
	if hymnHintRE.MatchString(line) {
		return nil
	}
	if regexp.MustCompile(`(?i)^Sermon\s*[:\-]`).MatchString(line) {
		return nil
	}
	if regexp.MustCompile(`(?i)^Special\s+Song\s*[:\-]`).MatchString(line) {
		return nil
	}
	if regexp.MustCompile(`(?i)^Theme(?:\s+Verse)?\s*[:\-]`).MatchString(line) {
		return nil
	}
	if regexp.MustCompile(`(?i)^(?:Verse\s+Reading|Memory\s+(?:Verse|Text)|Ayat\s+Bacaan)\s*[:\-]`).MatchString(line) {
		return nil
	}
	if regexp.MustCompile(`(?i)^(?:Family(?:\s*&\s*|\s+and\s+|/\s*)Youth|Family\s+of\s+the\s+Week|Youth\s+of\s+the\s+Week|Keluarga)`).MatchString(line) {
		return nil
	}
	if sectionRE.MatchString(line) && !strings.Contains(line, ":") {
		return nil
	}
	var m []string
	if b := bracketRole.FindStringSubmatch(line); b != nil {
		m = b
	} else if c := colonRole.FindStringSubmatch(line); c != nil {
		m = c
	} else {
		return nil
	}
	role := strings.TrimSpace(m[1])
	name := strings.TrimSpace(m[2])
	if role == "" || name == "" || clockRole.MatchString(role) {
		return nil
	}
	return &roleLine{Role: role, Name: name}
}

func withTiming(item Item, timing *string) Item {
	if timing == nil {
		return item
	}
	item.Timing = timing
	return item
}

func containsInt(xs []int, n int) bool {
	for _, x := range xs {
		if x == n {
			return true
		}
	}
	return false
}

func Normalize(parsed Rundown) Rundown {
	if parsed.Items == nil {
		parsed.Items = []Item{}
	}
	if parsed.UnmappedLines == nil {
		parsed.UnmappedLines = []string{}
	}
	if parsed.FailedHymnNumbers == nil {
		parsed.FailedHymnNumbers = []int{}
	}
	if parsed.FamilyPrayerRequest == nil && parsed.YouthPrayerRequest == nil && parsed.FamilyYouth != nil {
		parsed.FamilyPrayerRequest = parsed.FamilyYouth
	}
	return parsed
}
